corehq/form_processor/submission_post.py

```diff
@@ -219,12 +230,17 @@ def run(self):
         report_submission_usage(self.domain)
         failure_response = self._handle_basic_failure_modes()
         if failure_response:
-            return FormProcessingResult(failure_response, None, [], [], 'known_failures')
+            return FormProcessingResponse(failure_response, None, [], [], 'known_failures')
 
+        objects_to_track = []
+        if self.pre_processing_steps:
+            self.instance, objects_to_track, step_failure_response = self._pre_process_xform_xml()
+            if step_failure_response:
+                return step_failure_response
```

Comment:
I think we can reduce the footprint here quite a bit if we do something like this:
```python
context = self.get_submission_context()
with context:
    instance = context.get_form_instance()
    result = process_xform_xml(self.domain, instance, self.attachments, self.auth_context.to_json())
    context.post_process_form(result.submitted_form)
```

The the core HQ code we could have a `SubmissionContext` that does the current pre and post processing. The ICDS code can have it's own context class which adds in all the custom pre and post processing steps:

```python
class ICDSSubmissionContext(SubmissionContext):
    def __enter__(self):
        super().__enter__()
        # do pre-processing

    def post_process_form(self, xform):
        super().get_form_instance(xform)
        # do post-processing

```

Reply:
https://github.com/dimagi/commcare-hq/pull/28409/commits/3578ca3a5ac8239e474d9b77171b6c792577b2a1

That changed almost everything done so far!